lib/Dialect/FIRRTL/Transforms/DropConst.cpp

```diff
@@ -0,0 +1,184 @@
+      }
+    }
+
+    // Update the module signature with non-'const' ports
+    module->setAttr(FModuleLike::getPortTypesAttrName(),
+                    ArrayAttr::get(module.getContext(), portTypes));
+
+    if (!fmodule)
+      return result;
+
+    // Check the module body
+    visitDecl(fmodule);
+
+    // Drop 'const' from all registered values
+    for (auto [value, type] : constValuesToConvert)
+      value.setType(type);
+
+    return result;
+  }
+
+  void visitDecl(FModuleOp module) {
+    for (auto &op : llvm::make_early_inc_range(*module.getBodyBlock()))
+      dispatchVisitor(&op);
+  }
+
+  void visitExpr(ConstCastOp constCast) {
+    // Remove any `ConstCastOp`, replacing results with inputs
+    constCast.getResult().replaceAllUsesWith(constCast.getInput());
+    constCast->erase();
+  }
+
+  void visitStmt(WhenOp when) {
+    // Check that 'const' destinations are only connected to within
+    // 'const'-conditioned when blocks, unless the destination is local to the
+    // when block's scope
+
+    auto *previousNonConstConditionedBlock = nonConstConditionedBlock;
+    bool isWithinNonconstCondition = !when.getCondition().getType().isConst();
+
+    if (isWithinNonconstCondition)
+      nonConstConditionedBlock = &when.getThenBlock();
+    for (auto &op : llvm::make_early_inc_range(when.getThenBlock()))
+      dispatchVisitor(&op);
+
+    if (when.hasElseRegion()) {
+      if (isWithinNonconstCondition)
+        nonConstConditionedBlock = &when.getElseBlock();
+      for (auto &op : llvm::make_early_inc_range(when.getElseBlock()))
+        dispatchVisitor(&op);
+    }
+
+    nonConstConditionedBlock = previousNonConstConditionedBlock;
+  }
+
+  void visitStmt(ConnectOp connect) { handleConnect(connect); }
+
+  void visitStmt(StrictConnectOp connect) { handleConnect(connect); }
+
+  void visitUnhandledOp(Operation *op) {
+    // Register any 'const' results to drop 'const'
+    for (auto [index, result] : llvm::enumerate(op->getResults())) {
+      if (auto convertedType = convertType(result.getType()))
+        constValuesToConvert.push_back({result, convertedType});
+    }
+  }
+
+private:
+  /// Returns null type if no conversion is needed
+  Type convertType(Type type) {
+    if (auto base = type.dyn_cast<FIRRTLBaseType>()) {
+      return convertType(base);
+    }
+
+    if (auto refType = type.dyn_cast<RefType>()) {
+      if (auto converted = convertType(refType.getType()))
+        return RefType::get(converted.cast<FIRRTLBaseType>(),
+                            refType.getForceable());
+    }
+
+    return {};
+  }
+
+  /// Returns null type if no conversion is needed
+  FIRRTLBaseType convertType(FIRRTLBaseType type) {
+    auto nonConstType = type.getAllConstDroppedType();
+    return nonConstType != type ? nonConstType : FIRRTLBaseType{};
+  }
+
+  void handleConnect(FConnectLike connect) {
+    auto dest = connect.getDest();
+    auto destType = dest.getType();
```

Comment:
I think you need to break this down based on flow of bulk connects.  You won't correctly handle cases like the following:
```
firrtl.circuit "Foo" {
firrtl.module @Foo(in %cond: !firrtl.uint<1>, out %out: !firrtl.bundle<a flip: const.uint<1>>, in %in : !firrtl.bundle<a flip: uint<1>>) {
  firrtl.when %cond : !firrtl.uint<1> {
    firrtl.connect %out, %in : !firrtl.bundle<a flip: const.uint<1>>, !firrtl.bundle<a flip: uint<1>>
  }
}
}
```

Right now the verifier is rejecting these cases too, so I opened an issue here: https://github.com/llvm/circt/issues/5209

Reply:
Addressed in #5210

Reply:
Also updated to handle flips